Static Dialects Support

JVickery-TBS opened the pull request:
feat(dev): use static dialects;

- Started working on using the static dialects from the validation_options field.

Won't do much by itself, but offers that ability to use the dialects from the ckanext-validation plugin's `validation_options` field. https://github.com/open-data/ckanext-canada/pull/1430 <- this PR now sets static ones in an output validator. (https://github.com/open-data/ckanext-canada/pull/1430/commits/a63a46a5765476a3a4fd078385144b889e3fdd74)

Makes sense to use the `validation_options` field as the ckanext-validation uses goodtables/frictionless which just passes it all along to the Tabulator lib. And the ckanext-xloader also uses the Tabulator lib.

Still a work in progress as the subclassing for the Tabulator Streams and Parsers are tricky. Still trying to figure out how to get the `dialect` class property method to work...does not seem to be firing off. I imagine it is because the CSV parser method has some strange `__prepare_dialect` method that is not really using the `dialect` class method at all. So will have to most likely subclass that as well, and use `custom_parsers` argument to the `Stream` classes

JVickery-TBS committed:
feat(dev): use static dialects;

- Started working on using the static dialects from the validation_options field.

JVickery-TBS committed:
feat(dev): custom csv parser subclass;

- Added csv parser subclass.
- Added dialect sublcass for csv dialect.

JVickery-TBS committed:
feat(dev): custom csv dialect subclass;

- Syntax fix.

ThrawnCA commented on ckanext/xloader/loader.py:
```diff
@@ -35,33 +37,69 @@
+        """
+        if self.static_dialect:
+            if self.logger:
+                self.logger.info('Using Static Dialect for %s: %s', self.__format, dumps(self.static_dialect))
```
This will fire on every file load. Should it be at DEBUG level instead of INFO?

JVickery-TBS replied:
@ThrawnCA yeeeahhh haha, it will also fire off multiple times for a file possibly depending on the file encoding guessing thing.

I have yet to actually get this to work still. Just debugging it still to try and get the custom CSV parser to load. But does not seem to want to yet

JVickery-TBS replied:
Limited logging in the `load_csv` and `load_table` methods. Keeping `info` log level for our fork as it is what we desire for our setup

JVickery-TBS committed:
feat(dev): dialect class, finish mangling;

- Finalized mangle private method.
- Expanded Dialect subclass to work and verify that the dialect is valid.
- Limited dialect logging messages.

JVickery-TBS commented:
@wardi okay this seems to be working finally now. So if there is a dialect for the given format in a resource's `validation_options` field value, then it will pass that dict into the stream loader and parser classes.

As discussed, I had to use some private method mangling for the CSV parser. This code will only happen if there is a dialect in the `validation_options`, otherwise it will work as it did before. I had to do a Dialect class as well to have the name and class properties, this will also verify that the dialect is valid.

Granted the dialects should be valid for our setup as we are using the output validator for `validation_options` in: https://github.com/open-data/ckanext-canada/commit/a63a46a5765476a3a4fd078385144b889e3fdd74

JVickery-TBS committed:
Merge branch 'canada-v2.9' into feature/static-dialects

JVickery-TBS committed:
feat(dev): use static validation options;

- Use new static validation options dialect.
- Static encoding to come.

JVickery-TBS committed:
feat(dev): better logging, static encoding;

- Better logging for dialects.
- Add in options for static encoding.

JVickery-TBS committed:
fix(log): simplified logging;

- Just raw string it.

JVickery-TBS committed:
fix(dev): static encoding;

- Encoding parsing from validation options.

JVickery-TBS commented:
@wardi this now relies on this PR: https://github.com/open-data/ckanext-validation/pull/11

Granted, if it does not have that PR, it will just work as normal still. so does not fulllly rely on it

ThrawnCA commented on ckanext/xloader/jobs.py:
```diff
@@ -170,12 +170,16 @@ def _get_static_validation_options():
             mimetype=resource.get('format'),
             # (canada fork only): adds in dialect argument to pass static dialect
-            dialect=_get_static_validation_options().get('dialect', {})
+            dialect=static_options.get('dialect', {})
```
Doesn't this need a backslash to continue the line?

JVickery-TBS replied:
it's in the parenthesis of `load_csv`, so it is fine

wardi commented on ckanext/xloader/loader.py:
```diff
@@ -122,21 +162,32 @@ def _clear_datastore_resource(resource_id):
             file_format = mimetype.lower().split('/')[-1]
-            with UnknownEncodingStream(csv_filepath, file_format, decoding_result) as stream:
+            with UnknownEncodingStream(csv_filepath, file_format, decoding_result, dialect=dialect,
+                                       force_encoding=(True if encoding else False),
```
```suggestion
                                       force_encoding=bool(encoding),
```

wardi replied:
same change above, it's just a little cleaner

wardi commented:
this feels like a lot of code for a fairly small feature. See if the xloader maintainers have any suggestions for simplifying this

JVickery-TBS committed:
feat(dev): simplified syntax, resource field support;

- Simplified bool syntax.
- Added support for resource `validation_options` field if no static validation options.

JVickery-TBS commented:
@wardi bool syntax updated now, and I added in some code I forgot to add last week. It just adds support for a Resource's `validation_options` field value if there is no `static_validation_options` config value. Of course, we won't really be using that but just adding in the feature now as I will probably be throwing this upstream as well.

As for the a lot of code for a simple addition, Queenlands recently-ish made the parsing more robust and added in the encoding guessing to make Xloader a lot better. So I had to kind of code this stuff into that. Which was not a lot of code, just a couple parameters here and there.

The major issue with the code additions is the new classes. This is kind of an issue with Tabulator itself, because the Tabulator CSV Parser class has a different method which it uses to ALWAYS guess the CSV dialect. This parser works differently than all their other parsers. So even if we pass in the dialect values we want, it won't use them and will always call a private `__prepare_dialect` method (which only exists in this parser) AND the private method `__iter_extended_rows` calls it...ultimately it always does:
```
dialect = csv.Sniffer().sniff(separator.join(sample), delimiter)
```
And because both of those methods are private, that is why I had to mangle the `__prepare_dialect`

@ThrawnCA if you have any extra time, would love any feedback you have for this. Thanks!

JVickery-TBS committed:
fix(dev): child prop call;

- Fixed child prop call.

ThrawnCA commented on ckanext/xloader/jobs.py:
```diff
@@ -158,11 +158,31 @@ def xloader_data_into_datastore_(input, job_dict):
     logger.info('File hash: %s', file_hash)
     resource['hash'] = file_hash
 
+    # (canada fork only): use custom ckanext.validaion.static_validation_options
```
Misspelled "validation"

JVickery-TBS committed:
fix(typo): fixed typo;

- Fixed a typo, oops.

ThrawnCA commented:
> @ThrawnCA if you have any extra time, would love any feedback you have for this. Thanks!

IIRC the inflexibility of the parser class was a big part of why I replaced it wholesale with a custom post-processor.

Note that Tabulator is no longer maintained, so fixing it isn't really in the cards.
